### PR TITLE
perf(deepseek_v4): enable SGLANG_OPT_FP8_WO_A_GEMM on sm90 (Hopper)

### PR DESCRIPTION
**Pull request by guzekai01**
## Motivation

`SGLANG_OPT_FP8_WO_A_GEMM` runs the DeepSeek-V4 `wo_a` projection as an fp8
grouped matmul (`deep_gemm.fp8_einsum`). It was gated to Blackwell (sm100+)
because it assumed ue8m0-format scales. DeepGEMM's grouped fp8 GEMM also works
on Hopper (sm90) with fp32 scales, so this enables the path there. Opt-in;
sm100 behavior unchanged.

## Modifications

- `models/deepseek_v4.py`: pick scale format + recipe by `DEEPGEMM_SCALE_UE8M0`
  — sm100 ue8m0 `(1,1,128)`, sm90 fp32 `(1,128,128)` — in the `wo_a` forward,
  `_setup_fp8_wo_a_scales`, and layer init.
- `server_args.py`: keep the flag opt-in on sm90 (default off) instead of
  force-disabling below sm100; still force-off below sm90.
- `deep_gemm_wrapper/entrypoint.py`: skip the ue8m0 sanity check on the
  fp32-scale path.
- `test/manual/dsv4/wo_a_fp8_sm90_correctness.py`: correctness test (cosine vs
  the bf16 reference).

Validated on H20 (sm90), marlin MoE backend, A/B toggling only the flag:
DeepSeek-V4-**Flash** (TP8, single node) and DeepSeek-V4-**Pro** (TP8 + PP2,
2 nodes).

## Accuracy Tests

GSM8K, 1319 questions, parallel 100 (Invalid = 0.000 in all four runs):

| model | bf16 wo_a | fp8 wo_a |
| --- | ---: | ---: |
| Flash | 0.955 | 0.952 |
| Pro | 0.955 | 0.955 |

**Accuracy check — AIME2025**

Ran the AIME2025 regression on Hopper (sm90, H20) with the fp8 `wo_a` path enabled
(`SGLANG_OPT_FP8_WO_A_GEMM=1`, i.e. the sm90 opt-in this PR adds), Think-Max reasoning,
temperature 1.0 / top-p 1.0, on DeepSeek-V4-Flash:

| SGLANG_OPT_FP8_WO_A_GEMM | AIME2025 (opencompass, 30 Q) |
|---|---|
| `1` (fp8 wo_a, sm90 path) | **96.67%** |

In line with the DeepSeek-V4-Flash reference (~95%, repeated runs 93.3–100%), so the
sm90 fp8 `wo_a` GEMM introduces no accuracy regression.

## Speed Tests and Profiling

`bench_serving`. Δ = (fp8 − bf16)/bf16, negative = fp8 faster.

Prefill TTFT (ms, median; out=1, conc=1):

| ISL | Flash bf16 | Flash fp8 | Δ | Pro bf16 | Pro fp8 | Δ |
| ---: | ---: | ---: | ---: | ---: | ---: | ---: |
| 1k | 190.4 | 160.7 | −15.6% | 276.3 | 271.6 | −1.7% |
| 2k | 164.2 | 180.7 | +10.0% | 478.2 | 446.4 | −6.7% |
| 4k | 270.2 | 285.8 | +5.8% | 852.8 | 817.0 | −4.2% |
| 8k | 536.0 | 455.8 | −15.0% | 1474.2 | 1398.1 | −5.2% |
| 16k | 1092.3 | 984.0 | −9.9% | 2305.7 | 2242.5 | −2.7% |
| 32k | 2014.8 | 1886.7 | −6.4% | 4000.7 | 3858.7 | −3.5% |
| 64k | 3855.5 | 3821.0 | −0.9% | 7564.1 | 7275.4 | −3.8% |
| 128k | 8263.5 | 7856.5 | −4.9% | 14728.4 | 14566.0 | −1.1% |
| 256k | 17235.1 | 16642.7 | −3.4% | 30445.8 | 30208.7 | −0.8% |
| 512k | 38965.0 | 37682.9 | −3.3% | 65394.4 | 64566.9 | −1.3% |

Decode TPOT (ms; in=1024, out=1024, 3-round min):

| conc | Flash bf16 | Flash fp8 | Δ | Pro bf16 | Pro fp8 | Δ |
| ---: | ---: | ---: | ---: | ---: | ---: | ---: |
| 1 | 4.75 | 4.91 | +3.3% | 30.24 | 28.69 | −5.1% |
| 8 | 6.64 | 6.32 | −4.7% | 36.16 | 36.73 | +1.6% |
| 32 | 11.97 | 11.47 | −4.2% | 50.48 | 50.76 | +0.5% |
| 64 | 19.77 | 18.21 | −7.9% | 62.71 | 62.78 | +0.1% |

(Flash uses EAGLE; Pro under PP2 disables spec/overlap, so the two models'
absolute decode latencies are not directly comparable — read the per-model
bf16→fp8 Δ.)

## Memory

fp8 `wo_a` shrinks the per-rank weight; `mem-fraction-static` returns it to the
KV cache (`max_total_num_tokens`):

| model | bf16 weight (GB) | fp8 weight (GB) | bf16 max_total | fp8 max_total | Δ max_total |
| --- | ---: | ---: | ---: | ---: | ---: |
| Flash | 29.25 | 27.91 | 3,319,552 | 3,376,384 | +56,832 (+1.7%) |
| Pro | 70.64 | 68.77 | 5,252,608 | 5,426,688 | +174,080 (+3.3%) |

## Reproduction

Server (sm90, marlin MoE; `SGLANG_OPT_FP8_WO_A_GEMM=0` → bf16, `=1` → fp8):

```bash
# Flash — single node, TP8
SGLANG_OPT_FP8_WO_A_GEMM=0 python -m sglang.launch_server \
  --model-path <DeepSeek-V4-Flash> --trust-remote-code --host 0.0.0.0 --port 8001 \
  --tp-size 8 --moe-runner-backend marlin \
  --chunked-prefill-size 8192 --mem-fraction-static 0.8 \
  --max-running-requests 64 --cuda-graph-max-bs 64 \
  --speculative-algorithm EAGLE --speculative-num-steps 3 \
  --speculative-eagle-topk 1 --speculative-num-draft-tokens 4 \
  --enable-nsa-prefill-context-parallel --nsa-prefill-cp-mode round-robin-split \
  --tool-call-parser deepseekv4 --reasoning-parser deepseek-v4

# Pro — 2 nodes, TP8 + PP2 (PP disallows spec/overlap). Run on both nodes;
# node-rank 0 = master, node-rank 1 = worker.
SGLANG_OPT_FP8_WO_A_GEMM=0 python -m sglang.launch_server \
  --model-path <DeepSeek-V4-Pro> --trust-remote-code --host 0.0.0.0 --port 8001 \
  --tp-size 8 --pp-size 2 --nnodes 2 --node-rank <0|1> --dist-init-addr <master-ip>:20012 \
  --moe-runner-backend marlin --chunked-prefill-size 8192 --mem-fraction-static 0.95 \
  --max-running-requests 64 --cuda-graph-max-bs 64 --disable-overlap-schedule \
  --enable-nsa-prefill-context-parallel --nsa-prefill-cp-mode round-robin-split \
  --tool-call-parser deepseekv4 --reasoning-parser deepseek-v4
```

Client:

```bash
# accuracy (GSM8K)
python benchmark/gsm8k/bench_sglang.py --host http://localhost --port 8001 \
  --num-questions 1319 --parallel 100

# speed — prefill TTFT (e.g. ISL=8k, out=1)
python -m sglang.bench_serving --backend sglang --base-url http://localhost:8001 \
  --dataset-name random --random-input-len 8192 --random-output-len 1 \
  --num-prompts 8 --max-concurrency 1 --flush-cache --warmup-requests 3

# speed — decode TPOT (e.g. in=1024, out=1024, concurrency 32)
python -m sglang.bench_serving --backend sglang --base-url http://localhost:8001 \
  --dataset-name random --random-input-len 1024 --random-output-len 1024 \
  --num-prompts 32 --max-concurrency 32 --flush-cache --warmup-requests 3
```

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results (see above).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #29414600157](https://github.com/sgl-project/sglang/actions/runs/29414600157)
Latest PR Test (Extra): :white_check_mark: [Run #29414599754](https://github.com/sgl-project/sglang/actions/runs/29414599754)

### CODE REVIEW

File: python/sglang/srt/server_args.py [hunk 1]
```diff
@@ -6155,15 +6155,20 @@ def _handle_environment_variables(self):
...
         if is_cuda() and envs.SGLANG_OPT_FP8_WO_A_GEMM.get() and get_device_sm() < 100:
-            if envs.SGLANG_OPT_FP8_WO_A_GEMM.is_set():
-                logger.warning(
-                    "Disabling SGLANG_OPT_FP8_WO_A_GEMM: requires sm100+ (Blackwell), "
-                    "detected sm%d.",
-                    get_device_sm(),
-                )
-            envs.SGLANG_OPT_FP8_WO_A_GEMM.set(False)
+            if is_sm90_supported():
+                if not envs.SGLANG_OPT_FP8_WO_A_GEMM.is_set():
```
Comment (yuan-luo) [inline python/sglang/srt/server_args.py]: Please avoid nested if for so many layers. Moreover, SM90 opt-in only checks architecture, not deep_gemm_wrapper.ENABLE_JIT_DEEPGEMM. Later, MQALayer.forward() directly imports deep_gemm and calls fp8_einsum without fallback.

Comment (guzekai01) [reply python/sglang/srt/server_args.py]: Thanks for the review! All three addressed:

1. Flattened the nested `if` in `_handle_environment_variables` into a single
   `supported` predicate.
2. The gate now also requires `deep_gemm_wrapper.ENABLE_JIT_DEEPGEMM` (keyed on
   `DEEPGEMM_SCALE_UE8M0`, mirroring the runtime scale split), so the feature is
   disabled at startup whenever DeepGEMM JIT is unavailable.
3. Because of (2), `_FP8_WO_A_GEMM` is False when DeepGEMM is unavailable, so
   `MQALayer.forward()` never reaches `deep_gemm.fp8_einsum` without a backend —
   no per-request fallback needed.

One related find: DeepSeek-V4-Flash-FP8 ships `wo_a` as bf16 (no fp8 scale),
which would otherwise hit a cryptic `copy_with_check` downcast error. Added a
fail-fast in `load_weights` with an actionable message (use an fp8-`wo_a`
checkpoint, or set `SGLANG_OPT_FP8_WO_A_GEMM=0`).

PTAL, thanks!

**Review by yuan-luo — APPROVED** [review]
LGTM

**Comment by yuan-luo** [comment]
Please add AIME2025 result of DSv4 to make sure no accuracy drop.

**Comment by guzekai01** [comment]
> Please add AIME2025 result of DSv4 to make sure no accuracy drop.

Done.

File: python/sglang/srt/models/deepseek_v4.py [hunk 2]
```diff
@@ -78,6 +78,9 @@
 from sglang.srt.layers.logits_processor import LogitsProcessor
 from sglang.srt.layers.moe import get_moe_a2a_backend, should_use_dp_reduce_scatterv
 from sglang.srt.layers.moe.fused_moe_triton import FusedMoE
```
Comment (BBuf) [inline python/sglang/srt/models/deepseek_v4.py:87]: https://github.com/sgl-project/sglang/blob/main/python/sglang/srt/layers/quantization/fp8_utils.py#L10-L13

**Review by BBuf — APPROVED** [review]

File: test/manual/dsv4/wo_a_fp8_sm90_correctness.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+def main() -> None:
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--device", default="cuda")
+    parser.add_argument("--seed", type=int, default=0)
+    parser.add_argument("--cos-gate", type=float, default=0.999)
+    parser.add_argument("--decode-tokens", type=int, default=16)
+    args = parser.parse_args()
+
+    from sglang.srt.layers import deep_gemm_wrapper
```
Comment (BBuf) [inline test/manual/dsv4/wo_a_fp8_sm90_correctness.py]: Can we remove this line to the top of the file?

File: python/sglang/srt/models/deepseek_v4.py (hunk 2 above)
Comment (guzekai01) [reply python/sglang/srt/models/deepseek_v4.py:87]: Done

File: test/manual/dsv4/wo_a_fp8_sm90_correctness.py (hunk 3 above)
Comment (guzekai01) [reply test/manual/dsv4/wo_a_fp8_sm90_correctness.py]: Done
